File: OpenMP/solutions/omp_pi_localvars.f
      program pi

        use omp_lib
        implicit none

c       initialize variables
        integer :: i
        double precision :: pi_val = 0.d0
        double precision :: pi_val_local = 0.d0
        integer :: niter = 100000000

c       Get timing
        double precision :: start,end
        start=omp_get_wtime()

c       Calculate PI using Leibnitz sum
!$OMP PARALLEL PRIVATE(pi_val_local)
!$OMP DO
        do i=0,niter-1
          pi_val_local = pi_val_local + (-1)**i * (4 / (2*dble(i)+1))
        enddo
!$OMP ATOMIC
        pi_val = pi_val + pi_val_local
!$OMP END PARALLEL

c       Stop timing
        end=omp_get_wtime()-start;

c       Print result
        print *,"Pi estimate: ",pi_val," obtained in ",end," seconds"

      end program

File: OpenMP/solutions/omp_pi_localvars.c
#include <omp.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

int main (int argc, char *argv[])
{

  //initialize variables
  int i;
  double pi = 0;
  double pi_local = 0;
  int niter = 100000000;

  // Get timing
  double start,end;
  start=omp_get_wtime();

  // Calculate PI using Leibnitz sum
  /* Fork a team of threads */
#pragma omp parallel private(pi_local)
{
#pragma omp for
  for(i = 0; i < niter; i++)
  {
    pi_local = pi_local + pow(-1, i) * (4 / (2*((double) i)+1));
  }
#pragma omp atomic
  pi = pi + pi_local;
  
}

  // Stop timing
  end=omp_get_wtime();

  // Print result
  printf("Pi estimate: %.20f, obtained in %f seconds\n", pi, end-start);
}
